server/src/main/java/org/apache/druid/server/coordinator/duty/CoordinatorDutyUtils.java

```diff
@@ -0,0 +1,128 @@
+  {
+    int totalWorkerCapacity;
+    try {
+      final IndexingTotalWorkerCapacityInfo workerCapacityInfo =
+          FutureUtils.get(overlordClient.getTotalWorkerCapacity(), true);
+      totalWorkerCapacity = workerCapacityInfo.getMaximumCapacityWithAutoScale();
+      if (totalWorkerCapacity < 0) {
+        totalWorkerCapacity = workerCapacityInfo.getCurrentClusterCapacity();
+      }
+    }
+    catch (ExecutionException e) {
+      // Call to getTotalWorkerCapacity may fail during a rolling upgrade: API was added in 0.23.0.
+      if (e.getCause() instanceof HttpResponseException
+          && ((HttpResponseException) e.getCause()).getResponse().getStatus().equals(HttpResponseStatus.NOT_FOUND)) {
+        LOG.noStackTrace().warn(e, "Call to getTotalWorkerCapacity failed. Falling back to getWorkers.");
+        totalWorkerCapacity =
+            FutureUtils.getUnchecked(overlordClient.getWorkers(), true)
+                .stream()
+                .mapToInt(worker -> worker.getWorker().getCapacity())
+                .sum();
+      } else {
+        throw new RuntimeException(e.getCause());
+      }
+    }
+    catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    }
+
+    return totalWorkerCapacity;
+  }
+
+  /**
+   * Return the number of active tasks that match the task predicate provided. The number of active tasks returned
+   * may be an overestimate, as tasks that return status's with null types will be conservatively counted to match the
+   * predicate provided.
+   *
+   * @param overlordClient The overlord client to use to retrieve the list of active tasks.
+   * @param taskPredicate  The predicate to match against the list of retreived task status.
+   *
+   * @return the number of active tasks that match the task predicate provided
+   */
+  public static List<TaskStatusPlus> getNumActiveTaskSlots(
+      @Nonnull final OverlordClient overlordClient,
+      final Predicate<TaskStatusPlus> taskPredicate
+  )
+  {
+    final CloseableIterator<TaskStatusPlus> activeTasks =
+        FutureUtils.getUnchecked(overlordClient.taskStatuses(null, null, 0), true);
+    // Fetch currently running kill tasks
+    ImmutableList.Builder<TaskStatusPlus> taskStatuses = ImmutableList.builder();
+
+    try (final Closer closer = Closer.create()) {
+      closer.register(activeTasks);
+      while (activeTasks.hasNext()) {
+        final TaskStatusPlus status = activeTasks.next();
+
+        // taskType can be null if middleManagers are running with an older version. Here, we consevatively regard
+        // the tasks of the unknown taskType as the killTask. This is because it's important to not run
+        // killTasks more than the configured limit at any time which might impact to the ingestion
+        // performance.
+        if (status.getType() == null || (taskPredicate.apply(status))) {
```

Comment:
I observe something unclear in the `null` check here:
- First, we assume `status` is non-null because we deref status
- The, we call predicate; predicate first checks if `status` is null (when it can never be because of deref/NPE)
- Then we do comparison.

I think it would be good idea to clarify in Javadoc (and code) either one or the other situation
- `Predicate<TaskStatusPlus>` predicate "must be able to accept null", OR, predicate "will never be called with null".

Reply:
thanks, added here